lib/mayaUsd/fileio/primUpdaterManager.cpp

```diff
@@ -1724,4 +1747,270 @@ bool PrimUpdaterManager::readPullInformation(const MDagPath& dagPath, Ufe::Path&
+        TF_VERIFY(ancestorNode);
+        recursiveSetVisibility(ancestorNode, false);
+    } break;
+    case Ufe::SceneChanged::SubtreeInvalidate: {
+        // On subtree invalidate, the scene item itself has not had a structure
+        // change, but its children have changed.  There are two cases:
+        // - the node has children: from a variant switch, or from a payload
+        //   load.
+        // - the node has no children: from a payload unload.
+        // In the latter case, call recursiveHide(), because there is nothing
+        // below the invalidated node.
+
+        auto parentItem = Ufe::Hierarchy::createItem(op.path);
+        if (!TF_VERIFY(parentItem)) {
+            return;
+        }
+
+        auto parentHier = Ufe::Hierarchy::hierarchy(parentItem);
+        if (!parentHier->hasChildren()) {
+            auto ancestorNode = _pulledPrims.node(op.path);
+            if (ancestorNode) {
+                recursiveSetVisibility(ancestorNode, false);
+            }
+            return;
+        } else {
+            // On variant switch, given a pulled prim, the session layer will
+            // have path-based USD overs for pull information and active
+            // (false) for that prim in the session layer.  If a prim child
+            // brought in by variant switching has the same name as that of the
+            // pulled prim in a previous variant, the overs will apply to to
+            // the new prim, which would then get a path mapping, which is
+            // inappropriate.  Read children using the USD API, including
+            // inactive children (since pulled prims are inactivated), to
+            // support a variant switch to variant child with the same name.
+
+            auto parentUsdItem = std::dynamic_pointer_cast<MayaUsd::ufe::UsdSceneItem>(parentItem);
+            if (!TF_VERIFY(parentUsdItem)) {
+                return;
+            }
+            auto parentPrim = parentUsdItem->prim();
+            bool foundChild { false };
+            for (const auto& c :
+                 parentPrim.GetFilteredChildren(UsdPrimIsDefined && !UsdPrimIsAbstract)) {
+                auto cPath = parentItem->path().popSegment();
+                cPath = cPath
+                    + Ufe::PathSegment(
+                            c.GetPath().GetAsString(), MayaUsd::ufe::getUsdRunTimeId(), '/');
+
+                auto ancestorNode = _pulledPrims.node(cPath);
+                // If there is no ancestor node in the trie, this means that
+                // the new hierarchy is completely different from the one when
+                // the pull occurred, which means that the pulled object must
+                // stay hidden.
+                if (ancestorNode) {
+                    foundChild = true;
+                    recursiveSwitch(ancestorNode, cPath);
+                }
+            }
+            if (!foundChild) {
+                // Following a subtree invalidate, if none of the now-valid
```

Comment:
I wonder if a new variant could have some ancestor in the trie, just by happenstance because another pulled object happens to have the same ancestor? I'll be honest and I'm a bit foggy on what scenarios are possible or not.